docs/runtime-model.md

```diff
@@ -0,0 +1,356 @@
+notified in the future.
+
+This situation is generally resolved by adding a loop:
+
+```rust
+fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+    use self::BadTask::*;
+
+    loop {
+        match *self {
+            First(ref mut resource) => {
+                let value = try_ready!(resource.poll());
+                *self = Second(Resource2::new(value));
+            }
+            Second(ref mut resource) => {
+                try_ready!(resource.poll());
+                return Ok(Async::Ready(()));
+            }
+        }
+    }
+}
+```
+
+One way to think about it is that a task's `poll` function **must not**
+returned until it is unable to make any further progress due to its resources
+not being ready.
+
+Also note that, **functions that return `Async` must only be called from a
+task**. In other words, these functions only be called from code that has been
+submitted to `tokio::spawn` or other task spawn function.
+
+##### Yielding
+
+Somtimes a task must return `NotReady` without being blocked on a resource. This
+usually happens when computation to run is large and the task wants to return
+control to the executor to allow it to execute other futures.
+
+Yielding is done by notifying the current task and returning `NotReady`:
+
+```rust
+task::current().notify();
+return Ok(Async::NotReady);
+```
+
+Yield can be used to break up a CPU expensive computation:
+
+```rust
+struct Count {
+  remaining: usize,
+}
+
+impl Future for Count {
+    type Item = ();
+    type Error = ();
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        while self.count > 0 {
+            self.count -= 1;
+
+            // Yield every 10 iterations
+            if self.count % 10 == 0 {
+                task::current().notify();
+                return Ok(Async::NotReady);
+            }
+        }
+
+        Ok(Async::Ready(()))
+    }
+}
+```
+
+[contract]: https://docs.rs/futures/0.1.23/futures/future/trait.Future.html#tymethod.poll
+
+### Executors
+
+Executors are responsible for driving many tasks to completion. A task is
+spawned onto an executor, at which point the executor calls its `poll` function
+when needed. The executor hooks into the task system to receive resource
+readiness notifications.
+
+By decoupling the task system with the executor implementation, the specific
+execution and scheduling logic can be left to the executor implementation. Tokio
+provides two executor implementations, each with unique characteristics:
+[`current_thread`] and [`thread_pool`].
+
+When a task is first spawned onto the executor, the executor wraps it w/
+[`Spawn`][Spawn]. This binds the task logic with the task state (this is mostly
+required for legacy reasons). When the executor picks a task for execution, it
+calls [`Spawn::poll_future_notify`][poll_future_notify]. This function ensures
+that the task context is set to the thread-local variable such that
+[`task::current`][current] is able to read it. When calling
```

Comment:
Might be worth splitting the "When calling..." into another paragraph to break this up